azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -969,223 +930,336 @@ def do_retry():
+                        logger.debug(
+                            "{}({}): State changes CONNECTING -> CONNECTED. Connection established".format(
+                                self.name, event.name
+                            )
+                        )
+                        self.state = ReconnectState.CONNECTED
 
-            # UNEXPECTED DISCONNECTION (i.e. Connection has been lost)
-            elif self.state == ReconnectState.LOGICALLY_CONNECTED and self.pipeline_root.connected:
-                if self.pipeline_root.pipeline_configuration.connection_retry:
-                    # When we get disconnected, we try to reconnect as soon as we can.  We don't want
-                    # to reconnect right here because we're in a callback in the middle of being
-                    # disconnected and we want things to "settle down" a bit before reconnecting.
-                    #
-                    # We also use a timer to reconnect here because the "reconnect timer expired"
-                    # code path is well tested.  If we tried to immediately reconnect here, there
-                    # would be an entire set of scenarios that would need to be tested for
-                    # this case, and these tests would be identical to the "reconnect timer expired"
-                    # tests. Likewise, if there were 2 reconnect code paths (one immediate and one
-                    # delayed), then both those paths would need to be maintained as separate
-                    # flows
-                    logger.debug("{}({}): Attempting to reconnect".format(self.name, event.name))
-                    self.state = ReconnectState.WAITING_TO_RECONNECT
-                    self._start_reconnect_timer(0.01)
-                else:
-                    logger.debug(
-                        "{}({}): Not attempting to reconnect (Reconnect disabled)".format(
-                            self.name, event.name
+                    # EXPECTED CONNECTION (ReauthorizeConnectionOperation was previously issued)
+                    elif self.state is ReconnectState.REAUTHORIZING:
+                        logger.debug(
+                            "{}({}): State changes REAUTHORIZING -> CONNECTED. Connection re-established after reauthentication".format(
+                                self.name, event.name
+                            )
                         )
-                    )
-                    self.state = ReconnectState.LOGICALLY_DISCONNECTED
+                        self.state = ReconnectState.CONNECTED
 
-            # BAD STATE (this block should not be reached)
-            else:
-                logger.warning(
-                    "{}: DisconnectEvent received while in unexpected state - {}, Connected: {}".format(
-                        self.name, self.state, self.pipeline_root.connected
-                    )
-                )
+                    # BAD STATE (this block should not be reached)
```

Comment:
># BAD STATE (this block should not be reached) [](http://example.com/codeflow?start=20&length=46)

But what if it _is_ reached?  There's nothing preventing it.  You're right that it _shouldn't_ happen, but it might.

If you're in ReconnectState.CONNECTED and you get another ConnectedEvent, then this code happens to be right.  But what if you're in ReconnectState.DISCONNECTED and you get a ConnectedEvent?  Or DISCONNECTING?  Are you 100% sure that you're not connected even though Paho just told you that you are?

Reply:
As mentioned in your comment below, I have updated this logic to change the state anyway, but still raise the warning. This should minimize any interruption, but still bring an issue to our attention. If there is imprecision in our states, I want to know about it.

Reply:
>>  If there is imprecision in our states, I want to know about it, and figure out how we can get rid of it.

I couldn't agree more.  Who would have thought that "are we connected or not" would be such a pain in the butt to figure out.  I feel like maybe the best way to do this (not now, but maybe later) would be to have some object who's whole purpose is to keep track of connection state.  (I have something _like_ this in my track2 prototype, but it doesn't keep track of CONNECTING and DISCONNECTING, so it's not quite there yet.)